jgrapht-core/src/main/java/org/jgrapht/alg/tour/NearestInsertionHeuristicTSP.java

```diff
@@ -0,0 +1,326 @@
+     *
+     * @param subtour The vertices of the existing tour
+     * @param graph The graph
+     * @return List of vertices representing the complete tour
+     */
+    private List<V> augment(List<V> subtour, Graph<V, E> graph) {
+        Set<V> unvisited = new HashSet<>();
+        unvisited.addAll(graph.vertexSet());
+        unvisited.removeAll(subtour);
+        return augment(subtour, getClosest(subtour, unvisited, graph), unvisited, graph);
+    }
+
+    /**
+     * Augment an existing tour to give a complete tour
+     *
+     * @param subtour The vertices of the existing tour
+     * @param closestVertices Map of data for closest unvisited vertices
+     * @param unvisited Set of unvisited vertices
+     * @param graph The graph
+     * @return List of vertices representing the complete tour
+     */
+    private List<V> augment(List<V> subtour, Map<V, Closest<V>> closestVertices, Set<V> unvisited, Graph<V, E> graph) {
+        while (!unvisited.isEmpty()) {
+            // Select a city not in the subtour, having the shortest distance to any one of the cities in the subtoor.
+            Closest<V> closestVertex = chooseClosest(closestVertices);
+
+            // Determine the vertices either side of the selected tour vertex
+            int i = subtour.indexOf(closestVertex.getTourVertex());
+            V vertexBefore = subtour.get(i == 0 ? subtour.size() - 1 : i - 1);
+            V vertexAfter = subtour.get(i == subtour.size() - 1 ? 0 : i + 1);
+
+            // Find an edge in the subtour such that the cost of inserting the selected city between the edge’s cities will be minimal.
+            // Making assumption this is a neighbouring edge, test the edges before and after
+            double insertionCostBefore = graph.getEdgeWeight(graph.getEdge(vertexBefore, closestVertex.getUnvisitedVertex()))
+                    + closestVertex.getDistance()
+                    - graph.getEdgeWeight(graph.getEdge(vertexBefore, closestVertex.getTourVertex()));
+            double insertionCostAfter = graph.getEdgeWeight(graph.getEdge(vertexAfter, closestVertex.getUnvisitedVertex()))
+                    + closestVertex.getDistance()
+                    - graph.getEdgeWeight(graph.getEdge(vertexAfter, closestVertex.getTourVertex()));
+
+            // Add the selected vertex to the tour
+            if (insertionCostBefore < insertionCostAfter) {
+                subtour.add(i, closestVertex.getUnvisitedVertex());
+            } else {
+                subtour.add(i + 1, closestVertex.getUnvisitedVertex());
+            }
+
+            // Repeat until no more cities remain
+            updateClosest(closestVertices, closestVertex, unvisited, graph);
+        }
+        return subtour;
+    }
+
+    /**
+     * Class holding data for the closest unvisited vertex to a particular
+     * vertex in the tour.
+     *
+     * @param <V> vertex type
+     */
+    private static class Closest<V> implements Comparable<Closest<V>> {
```

Comment:
Does it make sense to create this class rather than to keep track of the closest edges instead?

Reply:
We need to sort by weight, and when we update the Map of closest vertices then we check the vertices in these objects, so in both cases we're avoiding going back to the Graph instance. So in my view this is an efficient way of doing it (whether the rest of the algorithm is efficient is another matter ;-) ).